v4.0.0 upgrade guide (#2164)
hashicorp/terraform-provider-vault: website/docs/guides/version_4_upgrade.html.markdown

## Why version 4.0.0?

We introduced version `4.0.0` of the Vault provider in order to make
performance improvements for deployments that manage many Vault secret or auth
engine mounts. This improvement required changes to the underlying Vault API
calls, which in turn would require policy adjustments in environments where
permissions are least privilege.

The change was deemed significant enough to warrant the major version bump.
In addition to the aforementioned performance improvements, all previously deprecated fields
and resources have been removed.

While you may see some small changes in your configurations as a result of
these changes, we don't expect you'll need to make any major refactorings.
However, please pay special attention to the changes noted in the TODO(JM) section.

## Which Terraform versions are supported?

Terraform versions `0.12.x` and greater are fully supported. Support for `0.11.x` has been removed.
If you are still on one of the `0.11.x` versions we recommend upgrading to the latest stable release of Terraform.

Please see the [Terraform Upgrade Guide](https://www.terraform.io/upgrade-guides/index.html)
for more info about upgrading Terraform.

> ## I accidentally upgraded to 4.0.0, how do I downgrade to `2.X`?

Reviewer: Is there a reason why we're providing a downgrade guide to 2.x but not 3.x?

Author: mistake, fixed

<!-- /TOC -->

## Provider Version Configuration

-> Before upgrading to version `4.0.0`, it is recommended to upgrade to the most
recent version of the provider (`3.25.0`) and ensure that your environment
successfully runs [`terraform plan`](https://www.terraform.io/docs/commands/plan.html)
without unexpected changes or deprecation notices.

It is recommended to use [version constraints](https://www.terraform.io/docs/configuration/providers.html#provider-versions)
when configuring Terraform providers. If you are following that recommendation,
update the version constraints in your Terraform configuration and run
[`terraform init`](https://www.terraform.io/docs/commands/init.html) to download
the new version.

If you aren't using version constraints, you can use `terraform init -upgrade`
in order to upgrade your provider to the latest released version.

For example, given this previous configuration:

```hcl
provider "vault" {
  # ... other configuration ...

  version = "~> 3.25.0"
}
```

An updated configuration:

```hcl
provider "vault" {
  # ... other configuration ...

  version = "~> 4.0.0"
}
```

## Provider Policy Changes

> 

Reviewer: Can we add a bit more detail what these policy changes mean (what does the arrow from one operation to another mean), and what action the user needs to take (e.g. update/add X to their token/role policy)?

Author: done

<!-- /TOC -->

## Provider Version Configuration

-> Before upgrading to version `4.0.0`, it is recommended to upgrade to the most
recent version of the provider (`3.25.0`) and ensure that your environment
successfully runs [`terraform plan`](https://www.terraform.io/docs/commands/plan.html)
without unexpected changes or deprecation notices.

It is recommended to use [version constraints](https://www.terraform.io/docs/configuration/providers.html#provider-versions)
when configuring Terraform providers. If you are following that recommendation,
update the version constraints in your Terraform configuration and run
[`terraform init`](https://www.terraform.io/docs/commands/init.html) to download
the new version.

If you aren't using version constraints, you can use `terraform init -upgrade`
in order to upgrade your provider to the latest released version.

For example, given this previous configuration:

```hcl
provider "vault" {
  # ... other configuration ...

  version = "~> 3.25.0"
}
```

An updated configuration:

```hcl
provider "vault" {
  # ... other configuration ...

  version = "~> 4.0.0"
}
```

## Provider Policy Changes

> Version `4.0.0` of the Vault provider made changes to the underlying Vault API

Author: The GET `sys/auth/:path` and GET `sys/mounts/:path` paths require Vault v1.10.0. Do you think we should call that out as well?

calls, which in turn may require policy adjustments in environments where
permissions are least privilege.

Please see the [Capabilities](https://developer.hashicorp.com/vault/docs/concepts/policies#capabilities)
section of the Vault Policies documentation for more information on Vault
policies.

### Auth method resource changes

The below table specifies what changed between version 3.X and 4.X for the
following resources:

#### Data sources
- `vault_auth_backend`

#### Resources
- `vault_auth_backend`
- `vault_gcp_auth_backend`
- `vault_github_auth_backend`
- `vault_jwt_auth_backend`
- `vault_ldap_auth_backend`
- `vault_okta_auth_backend`

<table>
<thead>
  <tr>
    <th colspan="2">3.X</th>
    <th colspan="2">4.X</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Method</td>
    <td>Path</td>
    <td>Method</td>
    <td>Path</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>sys/auth</td>
    <td>GET</td>
    <td>sys/auth/:path</td>
  </tr>
</tbody>
</table>

> For example, in version 3.X the following policy is required for the
> `vault_gcp_auth_backend` resource to perform the GET `sys/auth` HTTP operation:

Reviewer 2: We might want to say why these need to perform a GET on `sys/auth` (fetching mount metadata?) Just for illustration:

```suggestion
For example, in version 3.X the following policy is required for the
`vault_gcp_auth_backend` resource since its mount metadata is retrieved with the GET `sys/auth` HTTP operation:
```

Or since it's the same for all these resources we could state it in the intro for the section too.

```hcl
path "sys/auth"
{
    capabilities = ["read"]
}
```

> In version 4.X the following policy is required for the

Author: > is it meant to be all encompassing? Are there other specific Vault policy configuration that the user needs to be aware of

No, it is not meant to be all encompassing. There are other policies that the user would need to allow this resource to work properly. However, I decided to only highlight what was changed between 3.X and 4.X. There may be a better way to present that information though and I am open to suggestions.

> In version 4.X the following policy is required for the

Author: > I wonder if this sort of documentation should also appear on the main index page

That is something I considered as well. I think we should do that eventually. However, I believe it is out of scope for this PR.

> In version 4.X the following policy is required for the

Reviewer 3: I like scoping the upgrade guide to only include the changes between `3.X` and `4.X`, so I think we can be a bit more explicit about these policy changes not being all encompassing. I had two suggestions in that vein I thought might help:

- Would adding an extra line here at the end of [L154](https://github.com/hashicorp/terraform-provider-vault/pull/2164/files#diff-622e7159767cfaf13232d77fcbc29932cb285258f25ba6909da295670146e993R154) help? Something like `Note that the table below does not include any additional policies the individual resources might require.`
- Possible suggestion on tweaking the language to highlight that these policies are in addition to others:
```suggestion
In version 4.X the following policy must be included for the
```

`vault_gcp_auth_backend` resource to perform the GET `sys/auth/:path` HTTP operation:

```hcl
path "sys/auth/gcp"
{
    capabilities = ["read"]
}
```

### Secret engine resource changes

The below table specifies what changed between version 3.X and 4.X for the
following resources:

#### Resources
  - `vault_aws_secret_backend`
  - `vault_azure_secret_backend`
  - `vault_consul_secret_backend`
  - `vault_gcp_secret_backend`
  - `vault_mount`
  - `vault_pki_secret_backend_cert`
  - `vault_rabbitmq_secret_backend`
  - `vault_terraform_cloud_secret_backend`

<table>
<thead>
  <tr>
    <th colspan="2">3.X</th>
    <th colspan="2">4.X</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Method</td>
    <td>Path</td>
    <td>Method</td>
    <td>Path</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>sys/mounts</td>
    <td>GET</td>
    <td>sys/mounts/:path</td>
  </tr>
</tbody>
</table>

> For example, in version 3.X the following policy is required for the

Author: Yes, I would guess most users are using a highly privileged token. The assumptions I made when writing this guide were 
- the Upgrade Guide should be as concise as possible
- users running TFVP with tightly controlled token permissions already have an understanding of Vault Policies and TFVP
- they do not need guidance on TFVP's existing policy requirements
- they would be able to read the guide and see what they need to change

It may be that this view is too narrow and we should also cater to a wider audience, i.e. users that aren't using the principle of least privilege when running TFVP. I am open to suggestions here. :)

`vault_gcp_secret_backend` resource to perform the GET `sys/mounts` HTTP operation:

```hcl
path "sys/mounts"
{
    capabilities = ["read"]
}
```

In version 4.X the following policy is required for the
`vault_gcp_secret_backend` resource to perform the GET `sys/mounts/:path` HTTP operation:

```hcl
path "sys/mounts/gcp"
{
  capabilities = ["read"]
}
```

The below table specifies what changed between version 3.X and 4.X for the
following resources:

Resources
  - `vault_ad_secret_backend`

>   - `vault_nomad_secret_backend`

Reviewer 2: This seems like a new section? Or at least "Resources" should be a subheading?
